### PR TITLE
Get rid of unused parameter warnings

### PR DESCRIPTION
**Pull request by drwells**
This is a humongous (and dull) pull request: most of the unused parameters and variables in `deal.II` are appropriately marked (commented-out variable names, the `(void)` cast trick, etc) but this patch should catch the rest of them.

I updated the GCC flags (it also seems to work with clang) appropriately to include more warnings.

### CODE REVIEW

**Comment by tamiko** [comment]
What an effort. Thanks a lot!

I only have the one question stated above. Further, do you mind adding a small entry in <code>doc/news/changes.h</code>?

File: include/deal.II/lac/parallel_vector.templates.h [hunk 1]
```diff
@@ -308,6 +308,11 @@ namespace parallel
     Vector<Number>::compress_start (const unsigned int counter,
                                     ::dealii::VectorOperation::values operation)
     {
+      // These parameters are not used for some MPI and debug settings.
+      int ierr;
```
Comment (bangerth) [inline include/deal.II/lac/parallel_vector.templates.h]: What's the reason for pulling the declaration of ierr  out to the function level here and elsewhere?

**Comment by drwells** [comment]
@tamiko I did intend to add `-Wextra` only to debug mode. However, I (now that you mentioned it) see that it is a mistake to do this if we add `-Wno-unused` and `-Wno-unused-parameters` to the release mode flags. I will add it to release as well (in fact, it does not provide any extra warnings at the moment). I will add the entry to `doc/news/changes.h` once I hear a decision on what to do below.

@bangerth I moved `ierr` to silence a `-Wunused-variable` (or something similar) warning.You raise an interesting point: the commit where I `(void)` all those unused parameters is redundant because I disabled the related warnings in release mode previously.

Would it be better to disable unused parameter and variable warnings in release mode or to use the `(void)` trick? Maybe "both" is right to make non-gcc compilers happy? If the former is better, should I go back and remove all the extra `(void)`s? I know where they all are now :)

**Comment by tamiko** [comment]
> @tamiko I did intend to add `-Wextra` only to debug mode. However, I
> (now that you mentioned it) see that it is a mistake to do this if we
> add `-Wno-unused` and `-Wno-unused-parameters` to the release mode
> flags. I will add it to release as well (in fact, it does not provide
> any extra warnings at the moment). I will add the entry to
> `doc/news/changes.h` once I hear a decision on what to do below.

I just asked because you currently add "-Wectra" to DEAL_II_CXX_FLAGS
which is used for debug and release mode - which is already your
intented behavior. Can you move this addition up to the place where
"-pedantic" and "-Wall" is set up?

**Comment by bangerth** [comment]
> @bangerth <https://github.com/bangerth> I moved |ierr| to silence a
> |-Wunused-variable| (or something similar) warning.

I know, it was more about style. These `ierr` variables are used locally in 
just one line plus the check in the following line. I simply didn't see the 
point of pulling it out to function scope -- it separates declaration and use. 
If you need to `(void)` its use, do it right before or after the `Assert`. You 
may have to repeat this for each variable if there are multiple places in a 
function where we use an `ierr` variable, but I think it's still preferable to 
have multiple such variables than re-using a single one.


> Would it be better to disable unused parameter and variable warnings in
> release mode or to use the |(void)| trick? Maybe "both" is right to make
> non-gcc compilers happy? If the former is better, should I go back and remove
> all the extra |(void)|s? I know where they all are now :)

I have no preference either way. Your current approach seems fine to me. 
Certainly enabling the warning also in release mode helps us get more accurate 
warnings.

**Comment by drwells** [comment]
All right. I will rebase with the full set of warnings (and without `-Wno-unused-parameter`). @bangerth I agree with the previous style and will fix that as well.

Perhaps i should have made this point clearer: I got around writing most of the `(void)`s needed in `.cc` files with `-Wno-unused-parameter`, so there will be an additional (large) commit.

**Comment by drwells** [comment]
I believe that I have now resolved all of the outstanding issues. I force updated my own branch after rebasing with `fixup`.

File: doc/news/changes.h [hunk 2]
```diff
@@ -669,6 +669,12 @@ inconvenience this causes.
   <br>
   (Matthias Maier, 2014/12/29)
   </li>
+
+  <li> Improved: Used a variety of strategies to silence compiler warnings
+  about unused variables and unused parameters.
```
Comment (bangerth) [inline doc/news/changes.h]: Make it a complete sentence with a subject, e.g., "deal.II now uses a variety..."

**Comment by bangerth** [comment]
I'm ok with this patch if @tamiko is ok with the configuration bits.

The PR cannot be merged right now, probably because of the changes.h entry. Can you rebase and push again?

File: doc/news/changes.h (hunk 2 above)
Comment (drwells) [inline doc/news/changes.h]: Done.

File: cmake/setup_compiler_flags_gnu.cmake [hunk 3]
```diff
@@ -86,9 +87,9 @@ ENABLE_IF_SUPPORTED(DEAL_II_CXX_FLAGS "-Wno-deprecated-declarations")
 
 IF(CMAKE_CXX_COMPILER_ID MATCHES "Clang")
   #
-  # Silence Clang warnings about unused parameters:
+  # Silence Clang warnings about unused parameters (mainly caused by BOOST):
   #
-  SET(DEAL_II_CXX_FLAGS "-Qunused-arguments ${DEAL_II_CXX_FLAGS}")
```
Comment (tamiko) [inline cmake/setup_compiler_flags_gnu.cmake]: The <code>-Qunused-arguments</code> does something else and is highly necessary for older clang version due to a regression.

**Comment by tamiko** [comment]
Very nice. Let's rebase and merge.

**Comment by drwells** [comment]
You beat me to it! I have the rebase done. I should have checked this earlier :(

One more thing: I have a few more small commits that resulted from rebasing. It would be nice to get rid of them, but they are pretty minor. What do you recommend?

**Comment by tamiko** [comment]
> One more thing: I have a few more small commits that resulted from
> rebasing. It would be nice to get rid of them, but they are pretty
> minor. What do you recommend?

Can you cherry-pick them to master and open a new pull-request?

**Pull request closed without merging**